### DISCUSSION TITLE
Cleaning up Room On Disconnect

### DISCUSSION THREAD

**Discussion by danbergelt** [miguelgrinberg/Flask-SocketIO#2074] [Q&A] [answered]
Hello!

I have a question related to rooms when running with an external message queue (I.E. Redis) (and apologies if this question is better suited for the `python-socketio` repo).

When a client disconnects, are they automatically removed from the room? I couldn't find this happening from a quick scan of the code, and now I'm wondering if my Redis storage consumption will continually grow if I don't manually call `leave_room` in my `on_disconnect` handler. Would appreciate clarity on this (and what is recommended here in terms of cleanup work).

**Comment by miguelgrinberg** [comment, marked as answer]
Rooms are automatically garbage collected when they become empty. When a user disconnects they are removed from all the rooms they're in. Also nothing is stored on Redis, all storage is Python process memory. Redis is only used for sending notifications between processes.

The logic that removes a user from all the rooms is [here](https://github.com/miguelgrinberg/python-socketio/blob/main/src/socketio/base_manager.py#L87-L102) if you want to review it.

**Comment by danbergelt** [comment]
Thank you for the quick reply and the link to the code! This mostly clarifies things.

So just to confirm, if clients "A" and "B" are connected to process "Foo", and clients "C" and "D" are connected to process "Bar", and all clients are joined to room "Room", then a notification to room "Room" from process "Foo" will propagate to clients "C" and "D" by way of a message sent over the Redis pubsub channel? Since otherwise the process "Bar" would have no way of knowing about the notification.

**Reply by miguelgrinberg** [reply]
Yes. Both processes have a room "Room" each with their own subset of the clients that are in it and with no idea of what participants are in that room on the other hosts. The message queue is what allows a room operation to work across all hosts combined.

**Comment by danbergelt** [comment]
Thank you so much for your help (and for getting back so quickly).
